### PR TITLE
Integrate updated version of the package.

### PR DESCRIPTION
**Pull request by npnlee85**
This package was originally initiated by Scott Chamberlain in 2013. Not much work was done on it until a few years ago. Since then, the NPN has worked on a separate repo/fork improving this package and bringing it up-to-date with changes and improvements to our data services. As such, this pull request captures years of changes and is a complete overhaul of the package's functionality and documentation. The intention is that going forward changes can be pulled up to the ropensci repo more frequently and more discretely.

### CODE REVIEW

**Comment by sckott** [comment]
thanks, having a look

File: tests/test-all.R [hunk 1]
```diff
@@ -1,2 +1,2 @@
 library("testthat")
-test_check("rnpn")
+#test_check("rnpn")
```
Comment (sckott) [inline tests/test-all.R]: tests won't run on r check if you comment out this line

File: .gitignore [hunk 2]
```diff
@@ -4,4 +4,10 @@
 ._README.md
 .Rproj.user
 .Rhistory
-
+.RData
+inst/doc
+test-data
+test-data/
+test-data/*
```
Comment (sckott) [inline .gitignore]: you only need 1 of these 3 lines of test-data

File: DESCRIPTION [hunk 3]
```diff
@@ -4,24 +4,40 @@ Description: Programmatic interface to the
...
     events that occur at specific times.
-Version: 0.1.4.9100
+Version: 1.1.1
+Date: 2020-10-27
 License: MIT + file LICENSE
 Authors@R: c(
-    person("Scott", "Chamberlain", role = c("aut", "cre"), email = "myrmecocystus@gmail.com"),
-    person("Lee", "Marsh", role = "aut", email = "npnlee85@gmail.com")
+    person("Lee", "Marsh", role = c("aut", "cre"), email = "lee@usanpn.org"),
+    person("Scott", "Chamberlain", role = "aut", email = "myrmecocystus@gmail.com"),
```
Comment (sckott) [inline DESCRIPTION]: you can remove my email address

File: README.Rmd [hunk 4]
```diff
@@ -17,126 +22,87 @@ knitr::opts_chunk$set(
...
 [![codecov.io](https://codecov.io/github/ropensci/rnpn/coverage.svg?branch=master)](https://codecov.io/github/ropensci/rnpn?branch=master)
 
-`rnpn` is an R client for the US National Phenology Network API.
 
-National Phenology Network [API documentation](https://docs.google.com/document/d/1yNjupricKOAXn6tY1sI7-EwkcfwdGUZ7lxYv7fcPjO8/edit?hl=en_US).
+`rnpn` is an R client for interacting with the USA National Phenology Network data web services. These services include access to a rich set of observer-contributed, point-based phenology records as well as geospatial data products including gridded phenological model and climatological data.
+
+Documentation is available for the National Phenology Network [API documentation](https://docs.google.com/document/d/1yNjupricKOAXn6tY1sI7-EwkcfwdGUZ7lxYv7fcPjO8/edit?hl=en_US), which describes the full set of REST services this package wraps.
+
+There is no need for an API key to grab data from the National Phenology Network but users are required to self identify, on an honor system, against requests that may draw upon larger datasets.
```
Comment (sckott) [inline README.Rmd]: > users are required to self identify, on an honor system, against requests that may draw upon larger datasets.

how exactly do users do this?

File: DESCRIPTION [hunk 5]
```diff
@@ -4,24 +4,40 @@ Description: Programmatic interface to the
...
     testthat,
-    covr
+    covr,
+    knitr,
+    markdown,
+    rmarkdown,
+    RColorBrewer
+RoxygenNote: 7.1.0
+VignetteBuilder: knitr
 URL: https://docs.ropensci.org/rnpn (website)
```
Comment (sckott) [inline DESCRIPTION]: you need a trailing slash in the `https://docs.ropensci.org/rnpn` url - cran will complain about this if not there

File: R/npn_stations.R [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+
+
+#' Get number of stations by state.
+#'
+#' @export
+#' @template curl
+#' @return Number of stations by state as a data.frame.
+#' @examples \dontrun{
+#' head( npn_stationsbystate() )
```
Comment (sckott) [inline R/npn_stations.R]: here and other places, npn_stationsbystate is defunct, - update examples to make sure they work. you can run all examples via `devtools::run_examples(run = TRUE)`

**Comment by sckott** [comment]
a few things:

- vignettes: most maintainers are moving towards pre-built vignettes to avoid cran problems. the setup i like is having two files for each vignette, and use a make command (or do manually) to prebuild vignette locally -  e.g., https://github.com/ropensci/taxize/blob/master/Makefile#L4-L7 
- docs: we have pkgdown hosted docs for each ropensci package now - see https://docs.ropensci.org/rnpn/ - so you can remove docs folder. though you can keep pkgdown file as that is used
- at a later point you may want to switch to using github actions for CI since travis is going to cost money soon https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing
- tests take a long time to run. maybe use vcr https://github.com/ropensci/vcr for test caching. especially if you leave tests as is, use `skip_on_cran()` for all tests to avoid cran emails threatening to take the pkg down if any test fails
- code of conduct: we're now having pkg maintainers use the ropensci code of conduct - delete the conduct file here and add "* Please note that this package is released with a [Contributor Code of Conduct](https://ropensci.org/code-of-conduct/). By contributing to this project, you agree to abide by its terms." to the readme
- package imports: i'm not sure you're using all of the packages listed in imports/suggests. e.g., 

```r
# in R, in the rnpn package folder
map = pkgapi::map_package()
unique(vapply(sort(unique(map$calls$to)), function(w) strsplit(w,, split=":")[[1]][1], ""))
"base"       "curl"       "data.table" "httr"       "jsonlite"   "raster"     "rnpn"       "sp"         "tibble"     "utils"      "XML"
```

there could very well be good reasons for some of the packages you import, but some of them aren't used, e.g,. rgdal, tidyverse, magrittr, sf, sp

**Review by sckott — CHANGES_REQUESTED** [review]
a few things:

- vignettes: most maintainers are moving towards pre-built vignettes to avoid cran problems. the setup i like is having two files for each vignette, and use a make command (or do manually) to prebuild vignette locally -  e.g., https://github.com/ropensci/taxize/blob/master/Makefile#L4-L7 
- docs: we have pkgdown hosted docs for each ropensci package now - see https://docs.ropensci.org/rnpn/ - so you can remove docs folder. though you can keep pkgdown file as that is used
- at a later point you may want to switch to using github actions for CI since travis is going to cost money soon https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing
- tests take a long time to run. maybe use vcr https://github.com/ropensci/vcr for test caching. especially if you leave tests as is, use `skip_on_cran()` for all tests to avoid cran emails threatening to take the pkg down if any test fails
- code of conduct: we're now having pkg maintainers use the ropensci code of conduct - delete the conduct file here and add "* Please note that this package is released with a [Contributor Code of Conduct](https://ropensci.org/code-of-conduct/). By contributing to this project, you agree to abide by its terms." to the readme
- package imports: i'm not sure you're using all of the packages listed in imports/suggests. e.g., 

```r
# in R, in the rnpn package folder
map = pkgapi::map_package()
unique(vapply(sort(unique(map$calls$to)), function(w) strsplit(w,, split=":")[[1]][1], ""))
"base"       "curl"       "data.table" "httr"       "jsonlite"   "raster"     "rnpn"       "sp"         "tibble"     "utils"      "XML"
```

there could very well be good reasons for some of the packages you import, but some of them aren't used, e.g,. rgdal, tidyverse, magrittr, sf, sp

File: DESCRIPTION (hunk 3 above)
Comment (npnlee85) [reply DESCRIPTION]: Just your email or your name on the whole? You've been a guiding force and a mentor, seems appropriate to give credit where it's due!

Comment (sckott) [reply DESCRIPTION]: just remove the email address, keep my name :)

**Comment by npnlee85** [comment]
Still working through this but a few questions/comments:
 1) Re the docs folder: I'd like to keep that if it's not an issue, because we're using it for the Github pages associated with the forked repo: https://usa-npn.github.io/rnpn/  Maybe that's a little redundant? 
  2) Re the tests and CRAN: how quick is CRAN to pull a package? In many cases, it doesn't really make sense to cache responses from the server since in many ways, what's being tested is the response. On the other hand, I don't want the tests always taking forever, esp. on CRAN, invalidating test results, or throwing false negatives if a CRAN test is running when we coincidentally are having an outage. Thoughts on this?

**Comment by sckott** [comment]
sure, can keep docs folder, the main difference would be different urls of course, but styling will differ

you really don't want any failures in tests/examples on cran as they can quickly email after a failure and ask that it be fixed or the package be removed from cran within 2 weeks. id strongly suggest not running examples OR tests on cran that make real HTTP requests, b/c 1) their server may have a problem, 2) their internet may have a problem, 3) npn api/server may have a problem, etc. 

> what's being tested is the response.

Yes, but you're also testing the code that checks the response, munges the data into a data.frame, etc. What vcr does is remove the part where you're currently testing the http request/response behavior. 

I'd suggest not allowing real http requests on CRAN BUT definitely do real http requests on your CI system of choice and locally of course

**Ultimately its up to you** - just communicating my experience with cran

**Comment by npnlee85** [comment]
Currently attempting to incorporate vcr into the unit tests as discussed. However, I'm running into a problem wherein the particularly heavy requests aren't being cached (vcr generates an empty file). Per the documentation, vcr only works with httr and crul libraries but this part of the package is using curl instead; see this line - https://github.com/ropensci/rnpn/blob/2795dc7b81bdcc525ec7ea784b666b339348bb1e/R/npn_data_download.R#L671

I'm not sure if the problem is that I'm using curl, or because I'm streaming the response, OR if it's something else all together and wondering if you have any ideas @sckott . I could see changing this code to use crul or httr or I could hand mock vcr yml files but that sounds tedious.

**Comment by sckott** [comment]
ah sorry, vcr does not work with curl, sorry I should have seen that you were using curl and not recommended it. Yeah, if you switch to crul or httr you could use vcr.

**Comment by npnlee85** [comment]
Resolved most of the rest of the comments except for the bit about the vignettes. Honestly, not sure what to do with it, might need some guidance.  Rest of the comments should be sorted, submitted request for re-review.

File: DESCRIPTION [hunk 7]
```diff
@@ -4,24 +4,38 @@ Description: Programmatic interface to the
     Web Service methods provided by the National 'Phenology' Network
     (<https://usanpn.org/>), which includes data on various life history
     events that occur at specific times.
-Version: 0.1.4.9100
+Version: 1.1.1
+Date: 2020-10-27
```
Comment (sckott) [inline DESCRIPTION]: don't need to put the date field - that's filled in when you build the pkg

**Review by sckott — COMMENTED** [review]
just a few comments

- You recorded fixtures with `www-dev.usanpn.org` - if someone doesn't have that set that tries to run the tests they won't work - the most usage will be you and your coworkers and then whatever CI system, so maybe you just set the base url for the tests via maybe that `npn_set_env` function you have in the package. Actually I think the right base url gets set during R check but not if you're using R interactively. 
- Travis is changing soon to charge for open source repos soon, may want to switch to github actions or something else
- you should change names in license file

**Comment by npnlee85** [comment]
Putting in a new review request. A few comments to accompany that:
1) I believe this has been transitioned over correctly to using Github Actions. I've removed any references to travis, and I also removed appveyor simply because I didn't put it there, I don't use it. If there's value in having a secondary CI service or if it's standard for ropensci, I don't mind putting it back.
2) Re your comment about the text fixtures: I believe that's already in place. Each of the unit tests is prefaced with a call to npn_set_env to standardize the unit tests to use the dev environment which should match the test fixtures.

**Comment by sckott** [comment]
Thanks, I'll have a look

**Review by sckott — APPROVED** [review]

**Comment by sckott** [comment]
LGTM

**Pull request merged**